src/internal/connector/graph/concurrency_limiter.go

```diff
@@ -0,0 +1,52 @@
+package graph
+
+import (
+	"net/http"
+	"sync"
+
+	khttp "github.com/microsoft/kiota-http-go"
+)
+
+// ConcurrencyLimiter middleware limits the number of concurrent requests to graph API.
+// Currently we only enforce concurrency limit for exchange
+type ConcurrencyLimiter struct {
+	semaphoreExch chan struct{}
+}
+
+var (
+	once               sync.Once
+	concurrencyLimiter *ConcurrencyLimiter
+	// Outlooks expects max 4 concurrent requests
+	// https://learn.microsoft.com/en-us/graph/throttling-limits#outlook-service-limits
+	maxConcurrentRequestsExchange = 4
+)
+
+func InitializeConcurrencyLimiter(capacity int) {
+	once.Do(func() {
+		if capacity < 1 || capacity > maxConcurrentRequestsExchange {
+			capacity = maxConcurrentRequestsExchange
+		}
+
+		concurrencyLimiter = &ConcurrencyLimiter{
+			semaphoreExch: make(chan struct{}, capacity),
+		}
+	})
+}
+
+func GetConcurrencyLimiter() *ConcurrencyLimiter {
+	return concurrencyLimiter
+}
+
+func (cl *ConcurrencyLimiter) Intercept(
+	pipeline khttp.Pipeline,
+	middlewareIndex int,
+	req *http.Request,
+) (*http.Response, error) {
+	cl.semaphoreExch <- struct{}{}
```

Comment:
In the unlikely case that this gets called without cl getting initialized, this is going to panic.  You have two options: 1, if cl or cl.semaphore is nil, call `Initialize` here with a default concurrency value;  2, if cl or cl.semaphore is nil, return (nil, err).  I'd suggest 2, personally.

Reply:
True. I had ignored this because we only attach `ConcurrencyLimiter` if it's not nil. But I'll make this change.